kratos/input_output/file_logger_output.cpp

```diff
@@ -0,0 +1,48 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:         BSD License
+//                   Kratos default license: kratos/license.txt
+//
+//  Main authors:    Armin Geiser
+//
+//
+
+// System includes
+
+
+// External includes
+
+
+// Project includes
+#include "input_output/file_logger_output.h"
+
+
+namespace Kratos
+{
+    namespace {
+    std::ofstream* CreateNewLogFile(const std::string& rFileName)
+    {
+        // The file stream created here on the heap is deleted in the destructor
+        // of the FileLoggerOutput
+        return new std::ofstream(rFileName);
+    }
+    } // namespace
+
+    FileLoggerOutput::FileLoggerOutput(const std::string& rFileName)
+        : LoggerOutput(*CreateNewLogFile(rFileName))
```

Comment:
I did not find a better way to create the log file then this. Maybe someone has a better idea?

Reply:
I would use the `ofstream` constructor directly in the constructor of the `FileLoggerOutput`. Any reason that this doesn't work?

Reply:
It does not work because the `LoggerOutput` base class expects it as a constructor argument to store it as a member reference. And the constructor of the base class is called before the constructor of the derived class.

Reply:
I didn't expressed myself correctly. I was suggesting somthing like:

```c++
   FileLoggerOutput::FileLoggerOutput(const std::string& rFileName)
        : LoggerOutput(ofstream(rFileName))
```

Reply:
I see. That does not work because the ofstream is going out of scope immediately after the constructor of the base class and the object is destructed.. So the member reference becomes invalid.

Reply:
oops, just now I saw that output have a reference to the stream. 

So in this case I would change the `LoggerOutput::mrStream` to `unique_ptr<std::ostream> mpStream`. It would be cleaner in general as you explicitly giving the ownership to the base class.

Reply:
I hope i got it right, but: 

* If the LoggerOutput has ownership of the ostream, it will try to delete std::cout in the destructor in case of the default logger. I don't think that should be done.
* It will not be possible to give e.g. a stringstream to a logger and process it afterwards, because it will be deleted by the LoggerOutput destructor.

Reply:
You are absolutely right! I completely missed the entire image. 

So I would use a normal pointer in the base class (and even a set method) to have more flexibility in the future and ease your implementation.

Reply:
:+1: done